### DISCUSSION TITLE
Caching issue with shared albums

### DISCUSSION THREAD

**Discussion by CWempe** [photoprism/photoprism#2524] [Q&A]
I created a shared link for some albums/folders.

Over the time, new photos get uploaded to a folder and added to this shared link.

But I noticed the shared albums/links are not updated when the link had ben opened prevousely on the same device/browser.
It looks like the content is cached in the browser.

When I open the same link in an incognito window the correct content is shown.

Is this an actual issue, or am I dowing something wrong?
Is there an user-friendly workaround?

I tried restarting the container, but this resulted in "no albums found". (Android Chrome)
Again, it works in an incognito/private window.

I want to the the link with many guests of an event (next week) and want to upload new photos over the time.

**Comment by centralhardware** [comment]
is cntrl+F5 help? or delete cache in other ways in depend on your OS.

**Reply by CWempe** [reply]
Ctrl+F5 does not help! 😞 

Deleting the complete Browser Cache (tested in Firefox) did work. 👍 

But this is not a user-friendly solution, I think. :(

**Comment by CWempe** [comment]
Anybody got an idea, what I can do?

Just a naiv guess:
Is the progressive web app cached?
Could this be the issue?
Is there a way to prevent this from happening?

**Reply by lastzero** [reply]
It is generally possible to cache GET API requests, but we don't do that and these are the same endpoints for regular users and guests. So I don't see why caching would be any different. Before you dig deeper into our code, I recommend you check browser plugins, proxy servers and operating system settings, as some providers and even mobile carriers cache transparently without telling you.

**Reply by CWempe** [reply]
You are right.
It looks like an issue with my reverse proxy.

It seems to work when I access the service directly.

I am using https://nginxproxymanager.com/ and have trouble setting the `location /` settings.
Without the special settings everything else seemed to work so I did not thought of this.

I will look into this.

Thanks so far.

**Reply by CWempe** [reply]
I think I was too fast. 😞 

I can reproduce the issue without the reverse proxy.

* open shared link without issues in Firefox (no addons)
* restart container
* reload page (`Ctrl+F5` or `Ctrl+R`)
  *  no albums found
* clear browser cache
* reload page (`Ctrl+F5` or `Ctrl+R`)
  * now it works

![photoprism_caching_issue_local](https://user-images.githubusercontent.com/1270359/180078736-fd6f2b0f-626d-482c-a077-2013a1d2cc25.gif)

**Comment by lastzero** [comment]
The browser and the server store session information, so remembering albums is not unexpected as such. I thought the pictures in the albums don't update?

**Reply by CWempe** [reply]
No, newly added photos to an already shared album work.

But when I add a new album to an existing share-link, the new album is not shown if the user already opened the link in the past.

**Reply by lastzero** [reply]
Yes, this is to be expected, as the link token is only redeemed once for the albums that are shared at that time. Note that you can also redeem additional tokens (secrets) to get access to more albums.

**Reply by CWempe** [reply]
Just to clarify: The issue has nothing to do with "multiple albums".

I can break down the issue to this simple steps:

* share a single album
* open the shared link (http://192.168.1.179:2342/s/173shxbp6u not http://192.168.1.179:2342/s/173shxbp6u/arc8sca2bauei8ov) in an incognito browser window
  * see the shared album
* restart the container
* refresh the browser window
  * no album is shown

The following might be a separate issue:

When using the album-link (http://192.168.1.179:2342/s/173shxbp6u/arc8sca2bauei8ov) and pressing F5 after restarting the container I get redirected to http://192.168.1.179:2342/s/173shxbp6u/arc8sca2bauei8ov.
But opening the shared link again shows the photos of this album.
So this is not a big issue I think.

**Comment by CWempe** [comment]
I just checked again.

Newly added photos are shown as expected.

I also isolated the workaround: Just deleting cookies (not all browser data or cache) works to fix the issue.

So it looks like the issue is related to cookies somehow.

**Reply by CWempe** [reply]
If there is any "hack"  I can implement at the server side, I would appriciate it.

Maybe I can set the lifetime of the cookies to a few hours or so?

**Reply by lastzero** [reply]
We don't even use cookies, only localStorage:

- https://developer.mozilla.org/en-US/docs/Web/API/Window/localStorage

In addition, we use Websockets. So it might be helpful to use the JS console and the network request tab to check if the browser is for example still connected to the old instance (or uses old parameters). Maybe there is also some kind of JS error due to a failed request after the restart:

- https://docs.photoprism.app/getting-started/troubleshooting/browsers/

**Reply by CWempe** [reply]
There is nothing in the developer console when the issue occurs.

![image](https://user-images.githubusercontent.com/1270359/181828938-0e2496db-a6af-4284-86fd-00d43c757e4d.png)

It is irrelevant if there is only one or multiple albums in the shared link.

But I think there is not much more I can provide to help.

**Reply by lastzero** [reply]
See comment in the other thread.
